### ISSUE TITLE
It runs end-to-end after ~7 patches (cruise stub, numpy/moviepy/accelerate/peft pin shifts, flash-attn rms_norm stub, video transcode)

### ISSUE THREAD

**Issue by LukeLIN-web** [BriansIDP/video-SALMONN-o1#7]
# video-SALMONN-o1 Test Report

## Repo overview
video-SALMONN-o1 is the first open-source reasoning-enhanced audio-visual LLM for general video understanding (Sun et al., ICML 2025). It is a multi-modal model that consumes video frames + audio + text and produces step-by-step reasoning followed by a final answer. Architecture: Qwen2-7B language backbone + SigLIP-SO400M-patch14-384 vision encoder + Whisper-large-v3 audio encoder fronted by a speech Q-Former. Training uses SFT + a "process DPO" objective for step-level multimodal reward modeling. The companion benchmark, RivaBench (also called AVRBench in the paper), provides 4000+ expert-curated audio-visual MCQ across StandUp comedy, Academic presentations, and SynthDec (synthetic-video detection).

## Setup
- Clone: OK (https://github.com/BriansIDP/video-SALMONN-o1, 152 MB).
- Dependencies: PARTIAL — works after several deviations from README/pinned versions:
  - `torch==2.1.0+cu121`, `transformers==4.39.2` (per README) installed cleanly.
  - `accelerate==0.28.0` (per README) breaks: `peft.utils.loftq_utils` requires `accelerate.utils.memory.clear_device_cache`. Pinned `accelerate==0.29.3` + `peft==0.10.0` to resolve.
  - `numpy>=2` breaks torch 2.1 ABI — pinned `numpy<2`.
  - `moviepy==2.x` removed `moviepy.editor` — pinned `moviepy<2`.
  - `flash-attn` cannot build from source (uv build-isolation + pkg_resources issue); installed prebuilt wheel `flash_attn-2.5.6+cu122torch2.1cxx11abiFALSE`. That wheel does NOT ship the `dropout_layer_norm` CUDA extension, so `flash_attn.ops.rms_norm` is unimportable. Patched two files (`llava/model/modeling_qwen2.py`, `llava/model/multimodal_encoder/siglip_encoder_flash.py`) to make `rms_norm` import optional — it was imported but unused.
  - `cruise` (Bytedance internal HDFS/dist util) is not on PyPI. Three files import it without try/except (`modeling_qwen2.py`, `communication.py`, `attention.py`); the others already have try/except. Added a stub `DIST_ENV` (rank=0, world_size=1, seq-parallel-size=1) to each.
  - Misc: `einops`, `setuptools`, `wheel`, `ninja`, `peft`, `huggingface_hub`, `opencv-python`, `soundfile`, `pillow` installed as required by source imports (not in the README list).
- Model weights: `BrianatCambridge/video-SALMONN-o1` on HuggingFace (~17.4 GB across 4 safetensors shards) — downloaded and loaded successfully. Plus `google/siglip-so400m-patch14-384` (~3.3 GB) and `openai/whisper-large-v3` (~2.9 GB). Total local model footprint: ~23 GB. Combined repo + venv + weights + sample dataset = ~28 GB.
- Dataset (RivaBench): OK. Public at `https://huggingface.co/datasets/BrianatCambridge/RivaBench`. JSON files (`StandUp.json` 2128 entries, `Academic.json`, `synthdec.json` 200 entries) downloaded. Sample StandUp video+audio downloaded (`standupvideos/00001.mp4` + `standupaudios/00001.wav`). **Caveat**: the StandUp videos are AV1-encoded; the pinned `decord==0.6.0` cannot decode AV1 (`cannot find video stream with wanted index: -1`). Worked around by transcoding to H.264 with ffmpeg (`libx264`+`aac`). Decord then read the file fine.

## Inference test
- Command/script run:
  ```
  CUDA_VISIBLE_DEVICES=0 bash run_local_test.sh
  ```
  (`run_local_test.sh` is a copy of `run_inference.sh` with the BN/HDFS paths replaced by the local model + sample paths and `dataloader_num_workers=0`.)
- Input sample used: StandUp question `00001`
  - Video: `rivabench_sample/standupvideos/00001_h264.mp4` (~59 s, 1780 frames @ 30 fps after transcode)
  - Audio: `rivabench_sample/standupaudios/00001.wav`
  - Question: "Why did the audience find the comedian's statement, 'How do you know if someone is gay or if they're just from Dubai?' funny?"
  - Choices A-E (gold answer: **B** — "clever and unexpected comparison to highlight stereotypes and cultural misunderstandings")
- Output (actual model output, verbatim):
  ```
  Reasoning: The comedian makes a statement about how people in America are 'gay' but he is not, implying a cultural difference. <end_of_step>
  He then asks how someone can tell if someone is 'gay' or just 'from Dubai,' highlighting the stereotype that people from Dubai might be perceived as more conservative or less likely to be openly gay. <end_of_step>
  This clever comparison of two seemingly unrelated factors (sexual orientation and place of origin) creates a humorous contrast and plays on the audience's understanding of stereotypes. <end_of_step>
  Therefore, the answer is **B. The audience laughed because the comedian used a clever and unexpected comparison to highlight stereotypes and cultural misunderstandings.**<end_of_step><|im_end|>
  Answer: The final answer is:
  B. The audience laughed because the comedian used a clever and unexpected comparison to highlight stereotypes and cultural misunderstandings.<|im_end|>
  ```
- Time taken: ~22 s end-to-end per sample (`100%|██████████| 1/1 [00:22<00:00, 22.34s/it]`), plus ~5 s model load. The inference does two generation passes (CoT, then "final answer").
- GPU memory used: peak ~40.6 GiB on a single H100 PCIe (80 GB card, index 0). Model is loaded in bf16.

## Verdict
- **Can the code run? Partial-but-Yes.** It runs end-to-end after ~7 patches (cruise stub, numpy/moviepy/accelerate/peft pin shifts, flash-attn rms_norm stub, video transcode). None of these are deep — all are environment/packaging issues, not algorithmic blockers. Once patched, inference is fast and stable.
- **Is the dataset usable? Partial.** All JSONs + media files are publicly downloadable from HF. But the StandUp `.mp4`s are AV1-encoded and the project-pinned `decord==0.6.0` cannot decode them — you must transcode every video (or use a newer decord build with AV1 support, or swap in a different video reader). For a 4000-question full benchmark this is a non-trivial preprocessing step but mechanical.
- **Output quality (1 line): good** — the model correctly identified answer B and produced clean, step-tagged chain-of-thought matching the reasoning style described in the paper.
- **Blockers**:
  - `cruise` is not pip-installable (internal Bytedance pkg); minor stub fixes are needed in 3 files.
  - Prebuilt flash-attn wheels do not include the optional `dropout_layer_norm` CUDA extension that `flash_attn.ops.rms_norm` depends on; the source build would solve this but takes ~30 min on this machine and conflicts with uv's default build isolation. Two file patches make `rms_norm` (which is imported but never used) optional.
  - AV1-encoded videos in RivaBench's StandUp split are unreadable by pinned decord.
- **Recommendation: WORTH PURSUING.** The model loads, runs in bf16 on a single 80 GB H100 in ~22 s/sample, produces sensible step-by-step reasoning, and matches the gold answer on the one StandUp example tried. Plan for ~5 mostly-mechanical patches and a video-transcoding preprocessing step before a full RivaBench eval. With dataloader workers and a batched runner you can probably push throughput well below 22 s/sample.
